web-server/src/repositories/GenesisRepository.js

```diff
@@ -15,7 +20,24 @@ module.exports = {
             contract.tokenOfOwnerByIndex(address, idx)
         )
 
-        return (await Promise.all(tokenPromises)).map((t) => t.toNumber())
+        const ownedTokens = (await Promise.all(tokenPromises)).map((t) => t.toNumber())
+
+        // Number of Pixel Avatars tokens owned by the address
+        const avatarBalance = (await avatarContract.balanceOf(address)).toNumber()
+
+        // For each Pixel Avatars token the address owns we need to fetch the actual NFT
+        const avatarTokenPromises = [...Array(avatarBalance).keys()].map((idx) =>
+            avatarContract.tokenOfOwnerByIndex(address, idx)
+        )
+
+        const avatarOwnedTokens = (await Promise.all(avatarTokenPromises)).map((t) => t.toNumber())
+
+        return ownedTokens.map((tokenId) => {
+            return {
+                tokenId: tokenId,
+                minted: ( avatarOwnedTokens.indexOf(tokenId) > -1 )
+            }
+        })
```

Comment:
For this section, to reduce number of loops/maps what do you think about just having one loop to run through the tokens, and then just `await`s inside?

Something like this pseudo code:

```
const ownedTokens = {};

for (let idx=0; idx < avatarBalance; idx++) {
  const avatarToken = await avatarContract.tokenOfOwnerByIndex(address, idx);
  const avatarOwned = avatarOwnedTokens.indexOf(tokenId) > -1
}

return ownedTokens;
```

Reply:
Just a quick thought on this: Wouldn't that make the total duration longer since it would await for each token to finish before starting the next? :-)

Benefit of Promise.all is that they run in parallel and finishes as soon as the last one has finished :-)